velox/exec/OutputBuffer.cpp

```diff
@@ -640,4 +654,47 @@ bool OutputBuffer::isOverutilized() const {
   return (totalSize_ > maxSize_) && !atEnd_;
 }
 
+OutputBufferInfo OutputBuffer::getInfo() {
+  OutputBufferInfo info;
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    info.type = PartitionedOutputNode::kindString(kind_);
+    info.state = OutputBufferInfo::BufferState::OPEN;
+    if (noMoreBuffers_) {
+      info.state = OutputBufferInfo::BufferState::NO_MORE_BUFFERS;
+    }
+    if (atEnd_) {
+      info.state = OutputBufferInfo::BufferState::FLUSHING;
+    }
+    info.canAddBuffers = !noMoreBuffers_;
+    info.canAddPages = atEnd_;
+    if (bufferInfos_.size() < buffers_.size()) {
+      bufferInfos_.resize(buffers_.size());
+    }
+    for (auto i = 0; i < buffers_.size(); ++i) {
+      auto buffer = buffers_[i].get();
+      if (buffer) {
+        bufferInfos_[i] = buffer->getInfo();
+      }
+    }
+    info.buffers = bufferInfos_;
+  }
+  for (const auto& bufferInfo : info.buffers) {
+    info.totalBufferedBytes += bufferInfo.bufferedBytes;
+    info.totalBufferedRows += bufferInfo.bufferedRows;
+    info.totalBufferedPages += bufferInfo.bufferedPages;
+    info.totalBytesAdded += bufferInfo.bytesAdded;
+    info.totalRowsAdded += bufferInfo.rowsAdded;
+    info.totalPagesAdded += bufferInfo.pagesAdded;
+    info.totalBytesSent += bufferInfo.bytesSent;
+    info.totalRowsSent += bufferInfo.rowsSent;
+    info.totalPagesSent += bufferInfo.pagesSent;
+  }
+  if (info.state == OutputBufferInfo::BufferState::FLUSHING &&
+      info.totalBufferedBytes == 0) {
+    info.state = OutputBufferInfo::BufferState::FINISHED;
```

Comment:
A little confusing: if there is already data, then `totalBufferedBytes` will no longer be reduced to `0`, right? So the `FINISHED` state cannot be reached?

Reply:
Please read the newly added annotations for `xxxbufferedBytes`.